torch/_subclasses/meta_utils.py

```diff
@@ -463,7 +463,7 @@ def __call__(
             or (ignore_subclass and isinstance(t, torch.Tensor))
             or isinstance(t, FakeTensor)
         ):
-            if any(
+            if t.device.type != "xla" and any(
```

Comment:
Do you know why this was needed? (Or what codepath in the functionalization <> XLA integration hit this code?)

Reply:
Yep, AFAIK -- in dynamo codepath, pytorch converts the tensor into a fake tensor, and then run the ops on the fake tensor. And that process involves making a Meta tensor, and then turning that into the fake tensor subclass.

And now since XLA tensors are functional tensors, the `if` statement (`torch._is_functional_tensor(t)`) here returns true, which results in hitting the `return NotImplemented` code. So we added this check to bypass this if check.

More details at https://github.com/pytorch/xla/issues/4414#issuecomment-1380437595.

Reply:
So, a while ago we talked about the functionalization code paths for dynamo vs. lazy tensor. I think my take here is that it would be preferably for pytorch/XLA to go through all of this functionalization infra **only** when using the normal, non-dynamo code paths. And in the dynamo integration, for XLA to do what it used to do before this PR (not bother wrapping tensors into FunctionalWrappers, etc).

The main reason I'm leaning this way is because when you're using the torch.compile() / dynamo API, functionalization for XLA is completely redundant - our infra will send a graph to the backend to compile, and you're guaranteed that the graph is already functionalized.

If you end up keeping all of this functionalization logic in both code paths, one thing you might (?) run into is that there will be two levels of functionalization happening. This isn't really something that's tested / supported today. So my take here is that:

(1) If this one check here is enough to get everything working smoothly for dynamo, then this carve-out seems fine to me.

(2) If it's not, and you start hitting other weird functionalization failures, we might want to think about turning XLA functionalization off in the dynamo codepath.

Reply:
Thanks for the pointers. So far, this one check *was* enough to get everything working smoothly dynamo until recently we got a new dynamo regression for one of our unit tests (https://github.com/pytorch/xla/issues/4680). I didn't get chance to debug this too much in-depth, but outside from this it seems like we're not getting any other weird functionalization failures. 

@JackCaoG, what do you think about this? And if we were to modify PyTorch/XLA's dynamo to go off the functionalization codepath, would that be at our dynamo bridge at (https://github.com/pytorch/xla/blob/master/torch_xla/core/dynamo_bridge.py#L39 to manually update each tensor to a non-functional tensor?

Reply:
My take is: it looks like that we can keep this approach for now, and then investigate to skip functioanlization  for dynamo later on.

Reply:
We can try to disable the functionization in https://github.com/pytorch/xla/blob/master/torch_xla/core/dynamo_bridge.py#L224. That might be the right thing to do (If there is an easy flag to turn it on and off).

Reply:
Probably not, otherwise we can just land functionalization with the flag!!!

Reply:
well then there is no way for us to do that. The dynamo path's tracing will go through the same lazy tracing logic.

Reply:
Well, we can add a runtime flag... There are just a few places where we wrap the XLATensor into the FunctionalWrapper. I have been thinking about this for a while if we have to land the feature to unblock people...